---
title: "Ggplotting"
author: "Elisa Gomez de Lope"
date: '`r format(Sys.time(), "%d %B, %Y")`'
output: html_document
---

```{r echo=TRUE, message=TRUE, warning=TRUE, paged.print=TRUE}
library(dplyr)
library(vroom)
library(stringr)
library(ggplot2)
library(forcats)
```


> This practical aims at performing exploratory plots and how-to build layer by layer to be familiar with
the grammar of graphics. In the last part, a supplementary exercise will focus on plotting genome-wide CNV. 

##### Those kind of questions are optional {.bonus}

# Categorical data

We are going to use the built-in dataset called `ToothGrowth`. This dataset contains the teeth length of 60 guinea pigs which received 3 different doses of vitamin C (in mg/day), delivered either by orange juice (`OJ`) or ascorbic acid (`VC`). As for every (or near to every) function, most datasets shipped with a library contain also a useful help page (`?ToothGrowth`).

##### Is this dataset tidy?

The structure is tidy as it contains three variables as columns and 60 observations (rows). An ID besides the row index would be appreciated however


##### Plot the distributions as boxplots of the teeth lengths by the `dose` type received

```{block, opts.label = "tip"}
- `dose` needs to be a categorical data. You can convert it with `factor()`
- `geom_boxplot()` requires 2 aesthetics
    + `x` must be **categorical / discrete** (see first item)
    + `y` **must be continuous**
```

```{r}
ggplot(ToothGrowth) +
geom_boxplot(aes(y = len,
                 x = factor(dose)))
```


##### Attribute a filling colour to each delivery method (`supp` column)

```{r}
ggplot(ToothGrowth) +
geom_boxplot(aes(y = len,
                 x = factor(dose),
                 fill = supp))
```


When the dataset is tidy, it is easy to draw a plot telling us the story: vitamin C affects the teeth growth and the delivery method is only important for lower concentrations.

Boxplots are nice but misleading. The size of the dataset is not visible and the shapes of distributions could be better represented.

##### Use `violins` instead of `boxplots` to get a better view of the distribution shapes

```{block, opts.label = "tip"}
Set the option `trim = FALSE` to the `geom_violin()` to display distribution extremities
```

```{r}
ggplot(ToothGrowth) +
geom_violin(aes(y = len,
                 x = factor(dose),
                 fill = supp),
            trim = FALSE)

# Also, including points:
ggplot(ToothGrowth, aes(x = factor(dose),
                        y = len,
                        colour = supp)) +
  geom_violin(trim = FALSE) +
  geom_point(position = position_jitterdodge(dodge.width = 0.9))
```


Now we are missing summary values like the medians which are shown on boxplots. We should add them.

##### Add the median using the layer `geom_point(stat = "summary")` to the previous plot

```{block,  opts.label = "tip"}
Specify the `fun = "median"` and appropriate `size`, `colour` to get a big red dot representing the median

You can use a trick to get a quick legend without using strings in aesthetics. 
This will create a legend named with the string used. Concretely, `aes(colour = "median")`
can be used, and specifying a manual red color with `scale_color_manual(values = "red")` will 
enforce the red colour to the median dots.
```

```{r}
ggplot(ToothGrowth, aes(x = factor(dose),
                        y = len)) +
  geom_violin(trim = FALSE) +
  geom_point() +
  geom_point(stat = "summary", fun = "median", aes(colour = "median"), size = 4) + 
  scale_color_manual(values = "red") +
  labs(x = "vit C [mg/day]",
       y = "Tooth length",
       colour = NULL,
       title = "The Effect of Vitamin C on Tooth Growth in Guinea Pigs",
       caption = "C. I. Bliss (1952) The Statistics of Bioassay. Academic Press.") +
  theme_classic(base_size = 14) +
  theme(plot.caption.position = "plot",
        legend.position = "top")


ToothGrowth$dose <- as.factor(ToothGrowth$dose)
ggplot(ToothGrowth, aes(x=dose, y=len)) + 
  geom_violin(trim=FALSE) +
  geom_jitter(position = position_jitter(0.2)) +
  geom_point(stat = "summary", fun = "median", 
             aes(colour = "median"), size = 3)
```


Of note, a `ggplot` extension named `ggbeeswarm` proposes a very neat dotplot that fits the distribution.





# Scatter plots of penguins

The `penguins` dataset is provided by the `palmerpenguins` R package. As for every function, most data-sets shipped with a package contain also a useful help page (`?`).

##### If not done already, install the package `palmerpenguins` and load it.

```{r}
# install.packages("palmerpenguins")
library(palmerpenguins)

```

##### Plot the body mass on the _y_ axis and the bill length on the _x_ axis.

```{r}
penguins %>%
  ggplot(aes(x = bill_length_mm,
             y = body_mass_g)) +
  geom_point()
```


##### Plot again the body mass on the _y_ axis and the bill length on the _x_ axis, but with colour by `species`

```{r}
penguins %>%
  ggplot(aes(x = bill_length_mm,
             y = body_mass_g,
             colour = species)) +
  geom_point()
```




##### The `geom_smooth()` layer can be used to add a trend line. Try to overlay it to your scatter plot.

```{block, opts.label = "tip"}
by default `geom_smooth` is using a loess regression (< 1,000 points) and adds standard error intervals. 

- The `method` argument can be used to change the regression to a linear one: `method = "lm"`
- to disable the ribbon of standard errors, set `se = FALSE`

Be careful where the aesthetics are located, so the trend linear lines are **also** colored per species.
```

```{r}
penguins %>%
  ggplot(aes(x = bill_length_mm,
             y = body_mass_g,
             colour = species)) +
  geom_point() +
  geom_smooth(method = "lm" , formula = 'y ~ x', se = FALSE)

```



##### Adjust the aesthetics of point in order to

- the `shape` map to the originated `island`
- a fixed size of `3`
- a transparency of 40%

```{block,  opts.label = "tip"}
You should still have only 3 coloured linear trend lines. Otherwise check to which layer your are adding the aesthetic `shape`.
Remember that fixed parameters are to be defined outside `aes()`
```

```{r}
penguins %>%
  ggplot(aes(x = bill_length_mm,
             y = body_mass_g,
             colour = species)) +
  geom_point(aes(shape = island), size = 3, alpha = 0.6) +
  geom_smooth(method = "lm" , formula = 'y ~ x', se = FALSE)

```


##### Ajust the colour aesthetic to the `ggplot()` call to propagate it to both point and regression line.
Try the scale colour viridis for discrete scale (`scale_colour_viridis_d()`). Try to change the default theme to `theme_bw()`

```{r}
penguins %>%
  ggplot(aes(x = bill_length_mm,
             y = body_mass_g,
             colour = species)) +
  geom_point(aes(shape = island), size = 3, alpha = 0.6) +
  geom_smooth(method = "lm" , formula = 'y ~ x', se = FALSE) +
  scale_colour_viridis_d() 


```


##### Find a way to produce the following plot:


```{r}
penguins %>%
  ggplot(aes(x = bill_length_mm,
             y = body_mass_g,
             colour = species)) +
  geom_point(aes(shape = island), size = 3, alpha = 0.6) +
  geom_smooth(method = "lm" , formula = 'y ~ x', se = FALSE) +
  scale_y_continuous(labels = scales::comma) +
  scale_colour_viridis_d() +
  theme_bw(14) +
  labs(x = "Bill length (mm)",
       y = "Body mass (g)",
       color = "Penguin species",
       title = "Penguin bill length and body mass",
       subtitle = "Dimensions for male/female Adelie, Chainstrap and Gentoo Penguins \nat Palmer Station LTER",
       caption = "Horst AM, Hill AP, Gorman KB (2000)") +
  theme(plot.caption.position = "plot",
        legend.position = "right",
        plot.subtitle = element_text(size = 11))

```


```{block,  opts.label = "tip"}
Remember that:

- all aesthetics defined in the `ggplot(aes())` command will be inherited by all following layers
- `aes()` of individual geoms are specific (and overwrite the global definition if present).
- `labs()` controls of plot annotations
- `theme()` allows to tweak the plot like `theme(plot.caption = element_text(face = "italic"))` to render in italic the caption
```


# Supplementary exercises

##### Genome-wide copy number variants (CNV) detection {.bonus}

Let's have a look at a real output file for CNV detection. The used tool is called **Reference Coverage Profiles**: [RCP](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4330915/). 

It was developed by analyzing the depth of coverage in over 6000 high quality (>40Ã—) genomes.

In the end, for every _kb_ a state is assigned 
and similar states are merged eventually.

**state** means:

- 0, no coverage
- 1, deletion
- 2, expected diploidy
- 3, duplication
- 4, > 3 copies

## Reading data

The file is accessible [here](http://biostat2.uni.lu/practicals/data/CNV.seg). 
 `readr` can even read the file directly from the website so you don't need to download it locally.

`CNV.seg` has 5 columns and the first 10 lines look like:

![CNV](https://i.imgur.com/vzU4omM.png)

##### Load the file [`CNV.seg`](http://biostat2.uni.lu/practicals/data/CNV.seg) in R, assign to the name `cnv` {.bonus}

```{block, opts.label = "warning"}
several issues must be fixed:

- comment should be discarded.
- chromosome will be read as integers since first 1000 lines are **1**. But, X, Y are at the file's end.
- first and last column names are unclean. `#chrom` comtains a hash and `length (kb)`. Would be neater to fix this upfront.
```

```{r}
cnv <- vroom("http://biostat2.uni.lu/practicals/data/CNV.seg", # load file from url
      col_types = c("#chrom" = "c"), # change col type to character
      skip = 1L,  # skip first line (comment)
      col_select = list("chrom" = "#chrom", everything())) # remove the comment # from col name chrom
```


## Exploratory plots

##### Plot the counts of the different states. We expect a majority of diploid states.  {.bonus}

```{r}
cnv %>%
  ggplot(aes(x = state)) +
  geom_bar()
```


##### Plot the counts of the different states per chromosome. Might be worth freeing the **count** scale.  {.bonus}
 
```{r}
cnv %>%
  ggplot(aes(x = state)) +
  geom_bar() +
  facet_wrap(~chrom, scales = "free_y")
```


##### Using the previous plot, reorder the levels of chromosomes to let them appear in the karyotype order (1:22, X, Y)  {.bonus}

```{block,  opts.label = "tip"}
we could explicitly provide the full levels lists in the desired order. 

However, in the tibble, the chromosomes appear in the wanted order.

See the `fct_inorder()` function in the **forcats** package to take advantage of this.
```

```{r}
cnv %>%
  mutate(chrom = fct_inorder(chrom)) %>%
  ggplot(aes(x = state)) +
  geom_bar() +
  facet_wrap(~chrom, scales = "free_y")
```


##### Sexual chromosomes are not informative, collapse them into a **gonosomes** level  {.bonus}

```{block,  opts.label = "tip"}
See the `fct_collapse()` function in the **forcats**
```

```{r}
cnv %>%
  mutate(chrom = fct_collapse(chrom, 
               gonosomes = c("X", "Y")))

```


##### plot the genomic segments length per state

```{block,  opts.label = "tip"}
- The distributions are completely skewed: transform to log-scale to get a decent plot.
- Add the summary `mean` with a red dot and `median` with a purple dot using the **ToothGrowth** example
```
```{r}

cnv %>%
  mutate(length_kb = `length (kb)`) %>%
  select(-`length (kb)`) %>%
  ggplot(aes(x = factor(state),
             y = length_kb)) +
  geom_violin() +
  geom_point(stat = "summary", fun = "median", aes(colour = "median"), size = 15, shape = 95) + 
  geom_point(stat = "summary", fun = "mean", aes(colour = "mean"), size = 15, shape = 95) + 
  scale_color_manual(values = c(median = "orange", mean = "green")) +
  scale_y_log10() +
  annotation_logticks(sides = "l") +
  labs(x = "State",
       y = "Length (kb)") +
  theme_bw()
  
```


## Count gain / loss summarising events per chromosome

##### Filter the tibble only for *autosomes* and remove segments with no coverage and diploid (_i.e_ states 0 and 2 respectively). Save as `cnv_auto`.  {.bonus}

```{r}
cnv_auto <-cnv %>%
  filter((!state %in% c(0,2)) & (!chrom %in% c("X", "Y")))
  
```


##### We are left with state 1 and 3 and 4. Rename **1** as _loss_ and the others as _gain_ {.bonus}

##### Count the events per chromosome and per state  {.bonus}

##### For _loss_ counts, set them to negative so the barplot will be display up / down. Save as `cnv_auto_chr`   {.bonus}

```{r}
cnv_auto_chr <- cnv_auto %>%
  mutate(state = if_else(state == 1, "loss", "gain")) %>%
   count(chrom, state) %>%
   mutate(n = if_else(state == "loss", -n, n))
```


##### Plot `cnv_auto_chr` using the count as the `y` variable {.bonus}

```{r}
cnv_auto_chr %>%
  ggplot(aes(x = factor(chrom),
             y = n,
             fill = state)) +
  geom_col() +
  labs(x = NULL,
       y = "Count") +
  theme_classic() +
  theme(legend.position = c(1,1),
        legend.justification = c(2,1)) +
  scale_x_discrete(expand = c(0, 0)) +
  scale_y_continuous(labels = abs) +
  scale_fill_manual(values = c("springgreen3", "steelblue2"))
```


this is the final plot, where the following changes were made:

- labels of the `y` axis in absolute numbers
- set `expand = c(0, 0)` on the `x` axis. see [stackoverflow's answer](http://stackoverflow.com/a/22945857/1395352)
- use `theme_classic()`
- set the legend on the top right corner. Use a mix of `legend.position` and `legend.justification` in a `theme()` call.
- remove the label of the `x` axis, you could use _chromosomes_ if you prefer
- change the color of the fill argument with `c("springgreen3", "steelblue2")`


It is now obvious that we have mainly huge **deletions** on chromosome 10 and **amplifications** on chromosome 11.


In order to plot the genomic localisations of these events, we want to focus on the main chromosomes that were
affected by amplifications/deletions.

##### Lump the chromsomes by the number of CNV events (states 1, 3 or 4) keeping the 5 top ones and plot the counts {.bonus}


```{block,  opts.label = "tip"}
the function `fct_lump` from _forcats_ ease lumping. Just pick `n = 5` to get the top 5 chromosomes 
```

```{r}
cnv_auto %>%
  mutate(chrom = fct_lump(as.factor(chrom), n = 5)) %>%
  count(chrom) %>%
  ggplot(aes(x = chrom,
             y = n)) +
  geom_col() +
  theme_classic()
```

